File: core/CopyListWithRandomPointer/CopyListWithRandomPointer.java
package com.dbc;

import java.util.HashMap;
import java.util.Map;

public class CopyListWithRandomPointer {
    class Node {
        int val;
        Node next;
        Node random;

        public Node(int val) {
            this.val = val;
            this.next = null;
            this.random = null;
        }
    }

    public Node copyRandomList(Node head) {
        Node preHead = new Node(0);
        Node curNode = preHead, pointNode = head;
        Map<Node, Node> map = new HashMap<>();
        while (pointNode != null) {
            Node node = new Node(pointNode.val);
            curNode.next = node;
            curNode = curNode.next;
            map.put(pointNode, curNode);
            pointNode = pointNode.next;
        }
        for (Map.Entry<Node, Node> item : map.entrySet()) {
            if (item.getKey().random != null) {
                item.getValue().random = map.get(item.getKey().random);
            }
        }
        return preHead.next;
    }
}

File: core/CopyListWithRandomPointer/CopyListWithRandomPointer.py
from typing import Optional


class Node:
    def __init__(self, x: int, next: 'Node' = None, random: 'Node' = None):
        self.val = int(x)
        self.next = next
        self.random = random


def copyRandomList(head: 'Optional[Node]') -> 'Optional[Node]':
    new_head = Node(0)
    point_node, cur_node, node_map = head, new_head, {}
    while point_node:
        node = Node(point_node.val)
        cur_node.next = node
        cur_node = cur_node.next
        node_map[point_node] = cur_node
        point_node = point_node.next

    for old_node, new_node in node_map.items():
        if old_node.random:
            new_node.random = node_map[old_node.random]

    return new_head.next
